example/speech-demo/lstm_bucketing.py

```diff
@@ -0,0 +1,324 @@
+            label = labels[i].asnumpy().astype('int32')
+
+            mx.metric.check_label_shapes(label, pred_label)
+            
+            ind = np.nonzero(label.flat)
+            pred_label_real = pred_label.flat[ind]
+            #print label, pred_label, ind
+            label_real = label.flat[ind]
+            self.sum_metric += (pred_label_real == label_real).sum()
+            self.num_inst += len(pred_label_real)
+
+
+class BucketSentenceIter(mx.io.DataIter):
+    def __init__(self, train_sets, buckets, batch_size,
+            init_states, n_batch=None,
+            data_name='data', label_name='label'):
+
+        self.train_sets=train_sets
+        self.train_sets.initialize_read()
+           
+
+
+        self.data_name = data_name
+        self.label_name = label_name
+
+        buckets.sort()
+        self.buckets = buckets
+        self.data = [[] for k in buckets]
+        self.label = [[] for k in buckets]
+
+        self.default_bucket_key = buckets[0]
+
+        n = 0
+        while True:
+            (feats, tgts, utt_id) = self.train_sets.load_next_seq();
+            if utt_id == None:
+                break
+            if feats.shape[0] == 0:
+                continue
+            for i, bkt in enumerate(buckets):
+                if bkt >= feats.shape[0]:
+                    n = n + 1
+                    self.data[i].append((feats,tgts))
+                    break
+                        # we just ignore the sentence it is longer than the maximum
+            # bucket size here
+
+        # Get the size of each bucket, so that we could sample
+        # uniformly from the bucket
+        bucket_sizes = [len(x) for x in self.data]
+
+        print("Summary of dataset ==================")
+        for bkt, sz in zip(buckets, bucket_sizes):
+            print("bucket of len %3d : %d samples" % (bkt, sz))
+
+        bucket_size_tot = float(sum(bucket_sizes))
+        bucket_ratio_cum = [sum(bucket_sizes[:i+1]) / bucket_size_tot
+                for i in range(len(bucket_sizes))]
+
+        self.bucket_ratio_cum = bucket_ratio_cum
+        self.bucket_sizes = bucket_sizes
+        self.batch_size = batch_size
+        if n_batch is None:
+            n_batch = int(bucket_size_tot / batch_size)
+        self.n_batch = n_batch
+
+        self.init_states = init_states
+        self.init_state_arrays = [mx.nd.zeros(x[1]) for x in init_states]
+
+        self.provide_data = [('%s/%d' % (self.data_name, t), (self.batch_size,40))
+                for t in range(self.default_bucket_key)] + init_states
+        self.provide_label = [('%s/%d' % (self.label_name, t), (self.batch_size,))
+                for t in range(self.default_bucket_key)]
+
+    def __iter__(self):
+        init_state_names = [x[0] for x in self.init_states]
+        idx_bucket = np.arange(len(self.buckets))
+        np.random.shuffle(self.bucket_ratio_cum)
+        i = 0
+        for i_bucket in idx_bucket:
```

Comment:
It seems the data iteration consumes all the data in one bucket before moving to another bucket. I am not sure if having a full randomization might help the performance a bit. The recently updated PTB RNN example use full randomization. We might want to provide a common tool to help user doing this. But I think this can be postponed after GTC.